nova/src/ccs/mod.rs

```diff
@@ -0,0 +1,532 @@
+use ark_ec::{AdditiveGroup, CurveGroup};
+use ark_ff::Field;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use ark_spartan::polycommitments::PolyCommitmentScheme;
+
+use ark_std::Zero;
+use std::ops::Neg;
+
+#[cfg(feature = "parallel")]
+use rayon::iter::{
+    IndexedParallelIterator, IntoParallelIterator, IntoParallelRefIterator,
+    IntoParallelRefMutIterator, ParallelIterator,
+};
+
+use super::r1cs::R1CSShape;
+pub use super::sparse::{MatrixRef, SparseMatrix};
+use mle::vec_to_mle;
+
+pub mod mle;
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Error {
+    InvalidWitnessLength,
+    InvalidInputLength,
+    InvalidEvaluationPoint,
+    InvalidTargets,
+    NotSatisfied,
+}
+
+/// A type that holds the shape of the CCS matrices
+#[derive(Debug, Clone, PartialEq, Eq, CanonicalSerialize, CanonicalDeserialize)]
+pub struct CCSShape<G: CurveGroup> {
+    /// `m` in the CCS/HyperNova papers.
+    pub num_constraints: usize,
+    /// Witness length.
+    ///
+    /// `m - l - 1` in the CCS/HyperNova papers.
+    pub num_vars: usize,
+    /// Length of the public input `X`. It is expected to have a leading
+    /// `ScalarField::ONE` element, thus this field must be non-zero.
+    ///
+    /// `l + 1`, w.r.t. the CCS/HyperNova papers.
+    pub num_io: usize,
+    /// Number of matrices.
+    ///
+    /// `t` in the CCS/HyperNova papers.
+    pub num_matrices: usize,
+    /// Number of multisets.
+    ///
+    /// `q` in the CCS/HyperNova papers.
+    pub num_multisets: usize,
+    /// Max cardinality of the multisets.
+    ///
+    /// `d` in the CCS/HyperNova papers.
+    pub max_cardinality: usize,
+    /// Set of constraint matrices.
+    pub Ms: Vec<SparseMatrix<G::ScalarField>>,
+    /// Multisets of selector indices, each paired with a constant multiplier.
+    pub cSs: Vec<(G::ScalarField, Vec<usize>)>,
+}
+
```

Comment:
Putting the other type definitions (`CCSInstance` etc.) above the impls might improve readability.